pilot/pkg/networking/v1alpha3/gateway.go

```diff
@@ -16,161 +16,129 @@ package v1alpha3
 
 import (
 	"fmt"
-	"path"
-	"strconv"
-	"strings"
 
 	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	"github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
-	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
 	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
 	"go.uber.org/zap"
 
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+
 	networking "istio.io/api/networking/v1alpha3"
 	"istio.io/istio/pilot/pkg/model"
-	deprecated "istio.io/istio/pilot/pkg/proxy/envoy/v1"
 	"istio.io/istio/pkg/log"
 )
 
-func buildGatewayHTTPListeners(env model.Environment, node model.Proxy) ([]*xdsapi.Listener, error) {
+func buildGatewayListeners(env model.Environment, node model.Proxy) ([]*xdsapi.Listener, error) {
 
 	config := env.IstioConfigStore
 
-	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	// collect workload labels
+	workloadInstances, err := env.GetProxyServiceInstances(node)
 	if err != nil {
-		return nil, fmt.Errorf("listing gateways: %s", err)
+		log.Error("Failed to get gateway instances for router ", zap.String("node", node.ID), zap.Error(err))
+		return nil, err
+	}
+
+	var workloadLabels model.LabelsCollection
+	for _, w := range workloadInstances {
+		workloadLabels = append(workloadLabels, w.Labels)
 	}
 
+	gateways := config.Gateways(workloadLabels)
+
 	if len(gateways) == 0 {
 		log.Debug("no gateways for router", zap.String("node", node.ID))
 		return []*xdsapi.Listener{}, nil
 	}
 
-	gateway := &networking.Gateway{}
-	for _, spec := range gateways {
-		err := model.MergeGateways(gateway, spec.Spec.(*networking.Gateway))
-		if err != nil {
-			return nil, fmt.Errorf("merge gateways: %s", err)
-		}
+	// TODO: merging makes code simpler but we lose gateway names that are needed to determine
+	// the virtual services pinned to each gateway
+	//gateway := &networking.Gateway{}
+	//for _, spec := range gateways {
+	//	err := model.MergeGateways(gateway, spec.Spec.(*networking.Gateway))
+	//	if err != nil {
+	//		log.Errorf("Failed to merge gateway %s for router %s", spec.Name, node.ID)
+	//		return nil, fmt.Errorf("merge gateways: %s", err)
+	//	}
+	//}
+
+	// HACK for the above case
+	if len(gateways) > 1 {
+		log.Debug("Currently, Istio cannot bind multiple gateways to the same workload")
```

Comment:
Now this should actually be a warning, since we choose to program no gateways in this case.